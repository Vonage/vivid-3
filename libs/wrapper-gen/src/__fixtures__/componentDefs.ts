import { ComponentDef } from '../metadata/ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			forwardTo: {
				type: 'attribute',
				name: 'example-attribute',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			forwardTo: {
				type: 'attribute',
				name: 'type-union',
			},
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			forwardTo: {
				type: 'property',
				name: 'importedType',
			},
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			forwardTo: {
				type: 'attribute',
				name: 'value',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			forwardTo: {
				type: 'attribute',
				name: 'start',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			forwardTo: {
				type: 'property',
				name: 'forcedDomProp',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'boolean-attribute',
			forwardTo: {
				type: 'attribute',
				name: 'boolean-attribute',
				boolean: true,
			},
			type: [
				{
					text: 'boolean',
					vuePropType: 'Boolean',
				},
			],
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventNames: ['input'],
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'sameEvent',
			attributeName: 'example-attribute',
			eventNames: ['input'],
			valueMapping: '(event.target as any).exampleAttribute',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventNames: ['input:start'],
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
