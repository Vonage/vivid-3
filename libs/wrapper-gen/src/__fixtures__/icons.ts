import { IconsManifest } from '@vonage/vwc-consts';

export const iconsManifestStub: IconsManifest = [
	{
		id: '1-icon-1',
		keyword: ['icon-1'],
		tag: ['style_weight_solid', 'style_color_single', 'category_general'],
	},
	{
		id: 'icon-2',
		keyword: ['icon-2'],
		tag: ['style_weight_regular', 'style_color_multi', 'category_general'],
	},
];
