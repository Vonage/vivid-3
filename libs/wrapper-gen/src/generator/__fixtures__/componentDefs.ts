import { ComponentDef } from '../ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
			forceDomProp: true,
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventName: 'input',
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventName: 'input:start',
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
