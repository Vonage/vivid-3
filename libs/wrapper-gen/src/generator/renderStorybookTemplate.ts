import { ComponentDef } from './ComponentDef';
import { kebabToCamel, kebabToPascal } from './utils/casing';
import {
	isNumberLiteral,
	isStringLiteral,
	TypeUnion,
	withImportsResolved,
} from './types';

const isBasicString = (type: TypeUnion) =>
	type.length === 1 && type[0].text === 'string';
const isBasicNumber = (type: TypeUnion) =>
	type.length === 1 && type[0].text === 'number';
const isBasicBoolean = (type: TypeUnion) =>
	type.length === 1 && type[0].text === 'boolean';
const isLiteralUnion = (type: TypeUnion) =>
	type.every((t) => isStringLiteral(t.text) || isNumberLiteral(t.text));

const renderArgType = (attr: ComponentDef['attributes'][0]) => {
	const type = withImportsResolved(attr.type);

	if (isBasicString(type)) {
		return `{ type: { name: 'string', required: false }, control: 'text' }`;
	}
	if (isBasicNumber(type)) {
		return `{ type: { name: 'number', required: false }, control: { type: 'number' } }`;
	}
	if (isBasicBoolean(type)) {
		return `{ type: { name: 'boolean', required: false }, control: 'boolean' }`;
	}
	if (isLiteralUnion(type)) {
		return `{ type: { name: 'string', required: false }, control: 'select', options: [${type
			.map((t) => t.text)
			.join(', ')}]}`;
	}
	return `{ type: { name: 'object', required: false } }`;
};

export const renderStorybookTemplate = (def: ComponentDef) => {
	const propName = (attribute: string) => {
		const vueModel = def.vueModels.find((vm) => vm.attributeName === attribute);
		return vueModel ? vueModel.name : kebabToCamel(attribute);
	};
	const eventHandlerName = (event: string) =>
		`on${kebabToPascal(event.replaceAll(':', '-'))}`;

	const argTypesSrc = def.attributes
		.map((attr) => `${propName(attr.name)}: ${renderArgType(attr)},`)
		.join('\n');

	const events = [
		...def.events.map((event) => event.name),
		...def.vueModels.map((vm) => `update:${vm.name}`),
	];

	const eventsSrc = events
		.map((event) => `${eventHandlerName(event)}: { action: '${event}' },`)
		.join('\n');
	const eventListSrc = events
		.map((event) => `${eventHandlerName(event)}, `)
		.join('');
	const eventBindingsSrc = events
		.map((event) => `@${event}="${eventHandlerName(event)}"`)
		.join(' ');

	return `import { ${def.wrappedClassName} } from '@vonage/vivid-vue';

export const argTypes = {
  ${argTypesSrc}
  ${eventsSrc}
};

export const Template = args => ({
  components: { ${def.wrappedClassName} },
  setup() {
    const { ${eventListSrc} ...props } = args;
    return { ${eventListSrc} props };
  },
  template: '<${def.wrappedClassName} v-bind="props" ${eventBindingsSrc} />',
});
`;
};
