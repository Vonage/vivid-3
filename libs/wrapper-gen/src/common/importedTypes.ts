import { fetchIconsManifest, iconTypeFromManifest } from './icons';
import { Metadata } from './metadata';
import { makeTypeResolver } from './types';

export const makeImportedTypesResolver = async (metadata: Metadata) =>
	makeTypeResolver({
		IconId: iconTypeFromManifest(
			await fetchIconsManifest(metadata.iconsManifestUrl)
		),
	});
