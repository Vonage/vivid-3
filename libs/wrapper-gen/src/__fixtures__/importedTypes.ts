import { makeTypeResolver } from '../common/types';
import { iconTypeFromManifest } from '../common/icons';
import { iconsManifestStub } from './icons';

export const importedTypesResolverStub = makeTypeResolver({
	IconId: iconTypeFromManifest(iconsManifestStub),
});
