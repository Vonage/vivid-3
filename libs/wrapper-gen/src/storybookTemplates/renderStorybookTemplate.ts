import { ComponentDef } from '../common/ComponentDef';
import { kebabToPascal } from '../utils/casing';
import {
	isNumberLiteral,
	isStringLiteral,
	parseTypeStr,
	TypeResolver,
	TypeUnion,
} from '../common/types';
import { wrappedComponentName } from '../vueWrappers/name';

const isBasicString = (type: TypeUnion) =>
	type.length === 1 && type[0] === 'string';
const isBasicNumber = (type: TypeUnion) =>
	type.length === 1 && type[0] === 'number';
const isBasicBoolean = (type: TypeUnion) =>
	type.length === 1 && type[0] === 'boolean';
const isLiteralUnion = (type: TypeUnion) =>
	type.every((t) => isStringLiteral(t) || isNumberLiteral(t));

const renderArgType = (type: TypeUnion) => {
	if (isBasicString(type)) {
		return `{ type: { name: 'string', required: false }, control: 'text' }`;
	}
	if (isBasicNumber(type)) {
		return `{ type: { name: 'number', required: false }, control: { type: 'number' } }`;
	}
	if (isBasicBoolean(type)) {
		return `{ type: { name: 'boolean', required: false }, control: 'boolean' }`;
	}
	if (isLiteralUnion(type)) {
		return `{ type: { name: 'string', required: false }, control: 'select', options: [${type.join(
			', '
		)}]}`;
	}
	return `{ type: { name: 'object', required: false } }`;
};

export const renderStorybookTemplate = (
	def: ComponentDef,
	importedTypesResolver: TypeResolver
) => {
	const propName = (prop: string) => {
		const vueModel = def.vueModels.find((vm) => vm.propName === prop);
		return vueModel ? vueModel.name : prop;
	};
	const eventHandlerName = (event: string) =>
		`on${kebabToPascal(event.replaceAll(':', '-'))}`;

	const argTypesSrc = def.props
		.map(
			(prop) =>
				`${propName(prop.name)}: ${renderArgType(
					parseTypeStr(importedTypesResolver(prop.type))
				)},`
		)
		.join('\n');

	const events = [
		...def.events.map((event) => event.name),
		...def.vueModels.map((vm) => `update:${vm.name}`),
	];

	const eventsSrc = events
		.map((event) => `${eventHandlerName(event)}: { action: '${event}' },`)
		.join('\n');
	const eventListSrc = events
		.map((event) => `${eventHandlerName(event)}, `)
		.join('');
	const eventBindingsSrc = events
		.map((event) => `@${event}="${eventHandlerName(event)}"`)
		.join(' ');

	return `import { ${wrappedComponentName(def)} } from '@vonage/vivid-vue';

export const argTypes = {
  ${argTypesSrc}
  ${eventsSrc}
};

export const Template = args => ({
  components: { ${wrappedComponentName(def)} },
  setup() {
    const { ${eventListSrc} ...props } = args;
    return { ${eventListSrc} props };
  },
  template: '<${wrappedComponentName(
		def
	)} v-bind="props" ${eventBindingsSrc} />',
});
`;
};
