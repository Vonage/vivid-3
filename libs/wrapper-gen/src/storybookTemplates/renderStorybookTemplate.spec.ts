import { exampleComponent } from '../__fixtures__/componentDefs';
import { renderStorybookTemplate } from './renderStorybookTemplate';
import { importedTypesResolverStub } from '../__fixtures__/importedTypes';

describe('renderStorybookTemplate', () => {
	it('should render a storybook template', () => {
		expect(
			renderStorybookTemplate(exampleComponent, importedTypesResolverStub)
		).toMatchSnapshot();
	});
});
