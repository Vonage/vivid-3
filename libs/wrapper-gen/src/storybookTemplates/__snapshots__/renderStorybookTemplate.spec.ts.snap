// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderStorybookTemplate > should render a storybook template 1`] = `
"import { VExampleComponent } from '@vonage/vivid-vue';

export const argTypes = {
  sameEvent: { type: { name: 'string', required: false }, control: 'text' },
typeUnion: { type: { name: 'object', required: false } },
iconType: { type: { name: 'string', required: false }, control: 'select', options: ['1-icon-1', 'icon-2']},
modelValue: { type: { name: 'string', required: false }, control: 'text' },
start: { type: { name: 'string', required: false }, control: 'text' },
forcedDomProp: { type: { name: 'object', required: false } },
booleanProp: { type: { name: 'boolean', required: false }, control: 'boolean' },
  onExampleEvent: { action: 'example-event' },
onInput: { action: 'input' },
onInputStart: { action: 'input:start' },
onUpdateModelValue: { action: 'update:modelValue' },
onUpdateSameEvent: { action: 'update:sameEvent' },
onUpdateStart: { action: 'update:start' },
};

export const Template = args => ({
  components: { VExampleComponent },
  setup() {
    const { onExampleEvent, onInput, onInputStart, onUpdateModelValue, onUpdateSameEvent, onUpdateStart,  ...props } = args;
    return { onExampleEvent, onInput, onInputStart, onUpdateModelValue, onUpdateSameEvent, onUpdateStart,  props };
  },
  template: '<VExampleComponent v-bind="props" @example-event="onExampleEvent" @input="onInput" @input:start="onInputStart" @update:modelValue="onUpdateModelValue" @update:sameEvent="onUpdateSameEvent" @update:start="onUpdateStart" />',
});
"
`;
